taichi/backends/cuda/cuda_profiler.cpp

```diff
@@ -1,19 +1,167 @@
 #include "taichi/backends/cuda/cuda_profiler.h"
 #include "taichi/backends/cuda/cuda_driver.h"
 #include "taichi/backends/cuda/cuda_context.h"
-#include "taichi/system/timeline.h"
 
 TLANG_NAMESPACE_BEGIN
 
 #if defined(TI_WITH_CUDA)
 
-std::string KernelProfilerCUDA::title() const {
-  return "CUDA Profiler";
+bool check_device_capability() {
+  void *device;
+  int cc_major;
+  CUDADriver::get_instance().device_get(&device, 0);
+  CUDADriver::get_instance().device_get_attribute(
+      &cc_major, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR, device);
+  if (cc_major < 7) {
+    TI_WARN(
+        "CUPTI profiler APIs unsupported on Device with compute capability < "
+        "7.0 , fallback to default kernel profiler");
+    return false;
+  }
+  return true;
+}
+
+KernelProfilerCUDA::KernelProfilerCUDA() {
+// if Taichi was compiled with CUDA toolit, then use CUPTI
+// TODO : add set_mode() to select toolkit by user
+#if defined(TI_WITH_CUDA_TOOLKIT)
+  if (check_device_capability())
+    tool_ = ProfilingToolkit::cupti;
+#endif
+
+  if (tool_ == ProfilingToolkit::event) {
+    event_toolkit_ = std::make_unique<EventToolkit>();
+  } else if (tool_ == ProfilingToolkit::cupti) {
+    cupti_toolkit_ = std::make_unique<CuptiToolkit>();
+  }
+}
+
+// deprecated, move to trace()
+KernelProfilerBase::TaskHandle KernelProfilerCUDA::start_with_handle(
+    const std::string &kernel_name) {
+  TI_NOT_IMPLEMENTED;
+}
+
+void KernelProfilerCUDA::trace(KernelProfilerBase::TaskHandle &task_handle,
+                               const std::string &task_name) {
+  if (tool_ == ProfilingToolkit::event)
+    task_handle = event_toolkit_->start_with_handle(task_name);
+  else if (tool_ == ProfilingToolkit::cupti) {
+    TI_NOT_IMPLEMENTED;
+  }
+}
+
+void KernelProfilerCUDA::stop(KernelProfilerBase::TaskHandle handle) {
+  if (tool_ == ProfilingToolkit::event)
+    CUDADriver::get_instance().event_record(handle, 0);
```

Comment:
what happens to cupti in this case?

Reply:
`cuEvent` is controlled by host(CPU), needs to call star() and stop() explicitly.
With `CUPTI` , profiling is performed by device(GPU), just need to evaluate metric value form the CUPTI data image while `sync()`